llext: add a detached section test

lyakh opened the pull request:
Test LLEXT "detached section" functionality. This adds a test to load an extension with a function placed in an ad-hoc section and run a function in it.

lyakh committed:
llext: add a detached section test

Test LLEXT "detached section" functionality. This adds a test to load
an extension with a function placed in an ad-hoc section and run a
function in it.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh commented:
This doesn't work yet... debugging

lyakh committed:
llext: fix .bss on heap

With writable LLEXT but without pre-assigned section addresses
.bss section offset in its ELF header will not match .bss eventual
location as it's allocated on the heap. Use llext_loaded_sect_ptr()
to get a correct address in both cases.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 reviewed (commented):
Thanks for formalizing the test case! LGTM, I only have a string fix and a minor suggestion.

However, I may have missed something. How does this verify the `section_detached` machinery has done its job? It looks to me this would work fine even in the case it was completely ignored (due to a bug).

lyakh commented on tests/subsys/llext/simple/src/test_llext_simple.c:
```diff
@@ -437,6 +437,51 @@ ZTEST(llext, test_find_section)
+static bool test_section_detached(const elf_shdr_t *shdr)
+{
+	elf_shdr_t det_shdr;
+	int res = llext_get_section_header(detached_loader, detached_llext, ".detach", &det_shdr);
+
```
@pillo79 sorry, isn't this a matter of taste at most? ;-) I do prefer the concise form and I don't think Zephyr style dictates otherwise?

pillo79 replied:
Well the enforced coding style does require you to add spacing after the variables and I guess that is specifically to separate declarations from the function body. That extra line here also allows to group the function call with its assert just below, at a glance. 
But sure, I agree it's personal taste - it was a suggestion after all :slightly_smiling_face:

lyakh committed:
llext: move a calculation to a more logical location

Currently llext_link_plt() calculates offsets to relocation addresses
relative to the .text section and passes them to
arch_elf_relocate_global() and arch_elf_relocate_local(), where then
.text memory address is added to them. Instead it's more logical to
concentrate the entire calculation in the caller, which then also
removes the assumption, that all sections have the same VMA - LMA
offset from those functions.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: xtensa: (cosmetic) simplify a calculation

To obtain a section header reference we don't need to calculate its
location in the ELF image, we already have a pointer to all section
headers, just use them.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
llext: xtensa: fix RELATIVE relocations

With dynamically linked / shared ELF objects the displacement
between file offsets and memory addresses can differ between
sections. Therefore we cannot use .text for all such relocations and
have to locate the respective section instead.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

pillo79 commented on tests/subsys/llext/simple/src/test_llext_simple.c:
```diff
@@ -437,6 +437,66 @@ ZTEST(llext, test_find_section)
+#ifdef CONFIG_LLEXT_TYPE_ELF_SHAREDLIB
+	/*
+	 * Verify that the detached section is outside of the .text region.
+	 * This only works with the shared ELF type, because with other types
+	 * section addresses aren't "real," e.g. they can be 0.
```
Since this test is only performed in the `WRITABLE` case, you could probably avoid this issue by comparing `sh_offset`s instead? Those should be accurate in every kind of ELF file.

pillo79 reviewed (dismissed):
+1 to the Xtensa cleanups, and nice solution! I have a possible improvement to the test if you would like to lift the shared file test restriction.
